feat: Support spark transformers tracking with mlflow

serena-ruan opened the pull request:
## What changes are proposed in this pull request?

- Closes https://github.com/mlflow/mlflow/issues/5216
- Add support of pyspark.ml.Transformer to enable model tracking.

## How is this patch tested?

Tested local pytests and pack wheel packge to test on databricks.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Support all pyspark ml Transformers with the ability to save, log and load with mlflow.spark. 
https://github.com/mlflow/mlflow/issues/5216

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

serena-ruan committed:
feat: support Transformers for model tracking

Signed-off-by: Xinyue Ruan <serena.rxy@gmail.com>

serena-ruan committed:
Merge branch 'master' into serena/transformerSupport

akshaya-a commented on mlflow/spark.py:
```diff
@@ -451,9 +452,13 @@ def _save_model_metadata(
-        not isinstance(spark_model, PySparkModel)
+        (
+            not isinstance(spark_model, PySparkModel)
+            and not isinstance(spark_model, PySparkTransformer)
```
I'm a bit confused and my spark knowledge is very rusty at this point, but didn't Model inherit from Transformer? Can't we just move the check up a level in the type tree instead of checking either type?

dbczumar replied:
We could, but we might run into issues in the event of a breaking change to the inheritance hierarchy. However unlikely that may be, I like the robustness of the current approach to that case.

dbczumar reviewed (approved):
LGTM! Thanks @serena-ruan !

harupy commented on tests/spark/test_spark_model_export.py:
```diff
@@ -391,10 +393,39 @@ def test_sparkml_estimator_model_log(
-        sparkm.log_model(spark_model=spark_model_transformer.model, artifact_path="model0")
+@pytest.mark.parametrize("should_start_run", [False, True])
+@pytest.mark.parametrize("use_dfs_tmpdir", [False, True])
+def test_sparkml_transformer_model_log(
```
Do we need this test? It looks like `test_transformer_model_export` does the same thing.

harupy replied:
I think the `start_run` and `dfs_tmpdir` behaviors are already covered in other tests.

serena-ruan replied:
That makes sense, I'll delete this test

serena-ruan commented on tests/spark/test_spark_model_export.py:
```diff
@@ -21,7 +21,7 @@
 from mlflow import pyfunc
 from mlflow import spark as sparkm
-from mlflow.exceptions import MlflowException
+from mlflow.exceptions import MlflowException  # pylint: disable=unused-import
```
Sure, will do.

serena-ruan commented:
> https://github.com/mlflow/mlflow/blob/eb35d17c04a3a2c612a5b831c3d1accc8324054e/mlflow/spark.py#L730-L741
> 
> Can we fix line 740 as follows to allow using a transformer as a predictor?
> 
> ```python
>              for x in self.spark_model.setOutputCol("prediction").transform(spark_df).select("prediction").collect() 
> ```
> 
> This might be beyond the scope of this PR though.
> 
> cc @WeichenXu123

Yes I totally agree. For those transformers that inherit 'HasOutputCol' we can set it here to make the predict work for them by default.

harupy reviewed (commented):
LGTM, https://github.com/mlflow/mlflow/pull/5466#issuecomment-1061630285 is not a blocker, we can address it in a separate PR if necessary.

serena-ruan committed:
address comments and add support of predict for Transformers

Signed-off-by: Xinyue Ruan <serena.rxy@gmail.com>

serena-ruan commented:
> > https://github.com/mlflow/mlflow/blob/eb35d17c04a3a2c612a5b831c3d1accc8324054e/mlflow/spark.py#L730-L741
> > 
> > Can we fix line 740 as follows to allow using a transformer as a predictor?
> > ```python
> >              for x in self.spark_model.setOutputCol("prediction").transform(spark_df).select("prediction").collect() 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This might be beyond the scope of this PR though.
> > cc @WeichenXu123
> 
> Yes I totally agree. For those transformers that inherit 'HasOutputCol' we can set it here to make the predict work for them by default.

Just update on this cuz for PipelineModel (which is used to wrap around spark models) we don't have the outputCol param, so I changed the last stage in the PipelineModel to set outputCol as `prediction` in order to meet the requirement. Also updated the tests to test against the predict result for Transformer.

serena-ruan committed:
fix spark predict

Signed-off-by: Xinyue Ruan <serena.rxy@gmail.com>

harupy commented on mlflow/spark.py:
```diff
@@ -728,7 +734,13 @@ def predict(self, pandas_df):
+        from pyspark.ml import PipelineModel
+
         spark_df = self.spark.createDataFrame(pandas_df)
+        if isinstance(self.spark_model, PipelineModel):
+            # make sure predict work by default for Transformers
+            if self.spark_model.stages[-1].hasParam("outputCol"):
```
```suggestion
        if isinstance(self.spark_model, PipelineModel) and self.spark_model.stages[-1].hasParam("outputCol"):
            # make sure predict work by default for Transformers
```

can we merge the conditions?

serena-ruan committed:
Update mlflow/spark.py to merge conditions

Signed-off-by: Xinyue Ruan <serena.rxy@gmail.com>

harupy commented:
autoformat

mlflow-automation committed:
Autoformat: https://github.com/mlflow/mlflow/actions/runs/1956720120

Signed-off-by: mlflow-automation <mlflow-automation@users.noreply.github.com>

serena-ruan commented:
> autoformat

Ohh Thanks!! Love this automatic command!
